File: scheduling.pl
:- consult('helpers.pl').

generate_start_times_helper(Start, End, Duration, StartTimes) :-
    add_minutes(Start, Duration, EndTime),
    EndTime =< End,
    StartTimes = [Start | RestStartTimes],
    add_minutes(Start, 5, NextStart), % can modify time added for more granularity
    generate_start_times_helper(NextStart, End, Duration, RestStartTimes),
    !. % List is fully generated, stop backtracking
generate_start_times_helper(_, _, _, []).

generate_start_times(activity(_, StartRange, EndRange, Duration, _, _, _), StartTimes) :-
    generate_start_times_helper(StartRange, EndRange, Duration, StartTimes).

assign_start_time(activity(Name, StartRange, EndRange, Duration, _, _, _), activity(Name, StartRange, EndRange, Duration, _, ActualStart, ActualEnd)) :-
    generate_start_times(activity(Name, StartRange, EndRange, Duration, _, _, _), StartTimes),
    member(ActualStart, StartTimes),
    add_minutes(ActualStart, Duration, ActualEnd).

schedule(Activities, ScheduledActivities) :-
    schedule_helper(Activities, [], ScheduledActivities).

schedule_helper([], ScheduledActivities, ScheduledActivities).
schedule_helper([Activity|Rest], PreviousActivities, ScheduledActivities) :-
    assign_start_time(Activity, ScheduledActivity),
    conflict_check(ScheduledActivity, PreviousActivities),
    append([ScheduledActivity], PreviousActivities, UpdatedPreviousActivities),
    schedule_helper(Rest, UpdatedPreviousActivities, ScheduledActivities).

conflict_check(_, []).
conflict_check(Activity, [OtherActivity|Rest]) :-
    \+ conflict_found(Activity, OtherActivity),
    conflict_check(Activity, Rest).

conflict_found(activity(_, _, _, _, _, Start1, End1), activity(_, _, _, _, _, Start2, End2)) :-
    Start1 < End2,
    End1 > Start2.

File: HaskellCurrent/scheduling.hs
module Scheduling where

import Types (Activity(..))
import Helpers (addMinutes)
import Data.List (find, sortBy, sortOn)
import Data.Maybe (fromJust)
import MyIO (printActivities)

-- import Myio (printActivities)


-- Generates all possible start times for an activity
generateStartTimes :: Activity -> [Int]
generateStartTimes activity =
  generateStartTimesHelper (startRange activity) (endRange activity) (duration activity)
  where
    generateStartTimesHelper start end duration
      | addMinutes start duration <= end = start : generateStartTimesHelper (addMinutes start 1) end duration
      | otherwise = []

-- Chooses a value from the start time options list, based on if there is a conflict with a previous activity
chooseStartTime :: [Int] -> [Activity] -> [Activity] -> Int -> Maybe Int
chooseStartTime [] _ _ _ = Nothing
chooseStartTime (startTime:xs) previousActivities remainingActivities duration =
  let endTime = addMinutes startTime duration
      ifConflictFound = conflictsWith startTime endTime (previousActivities ++ remainingActivities)
  in if ifConflictFound then chooseStartTime xs previousActivities remainingActivities duration else Just startTime



conflictsWith :: Int -> Int -> [Activity] -> Bool
conflictsWith start end activities =
  let conflictingActivities = filter (conflictFound start end) activities
  in not $ null conflictingActivities

-- Check if two activities have a conflict
conflictFound :: Int -> Int -> Activity -> Bool
conflictFound start1 end1 activity2 =
  let start2 = actualStart activity2
      end2 = actualEnd activity2
  in start1 < end2 && end1 > start2


schedule :: [Activity] -> [Activity]
schedule activities = head $ scheduleActivities activities []

scheduleActivities :: [Activity] -> [Activity] -> [[Activity]]
scheduleActivities [] scheduledActivities = [scheduledActivities]
scheduleActivities (currentActivity:remainingActivities) scheduledActivities =
  [ scheduledActivity : restOfSchedule
  | startTime <- generateStartTimes currentActivity
  , not $ conflictsWith startTime (addMinutes startTime (duration currentActivity)) scheduledActivities
  , let scheduledActivity = currentActivity { actualStart = startTime, actualEnd = addMinutes startTime (duration currentActivity) }
  , restOfSchedule <- scheduleActivities remainingActivities (scheduledActivities ++ [scheduledActivity])
  ]



-- testing
main :: IO ()
main = do
  let gym = Activity "gym" 830 1130 120 True 0 0
      lunch = Activity "lunch" 1200 1300 30 True 0 0
      homework = Activity "homework" 1200 1930 120 True 0 0
      breakfast = Activity "breakfast" 830 850 20 True 0 0
      ponderLife = Activity "ponder life" 830 1630 120 True 0 0
      kickboxing = Activity "kickboxing" 1200 1500 120 True 0 0

      scheduledActivities = schedule [gym, lunch, homework, breakfast, ponderLife, kickboxing]
      
  putStrLn "Scheduled Activities: "
  printActivities scheduledActivities
